reconcile/acs_rbac.py

```diff
@@ -0,0 +1,558 @@
+
+        :param query_func: function which queries GQL server
+        :return: list of AcsRole derived from oidc-permission-1 definitions
+        """
+
+        query_results = acs_rbac_query(query_func=query_func).acs_rbacs
+        if query_results is None:
+            return []
+
+        permission_usernames: dict[Permission, list[str]] = defaultdict(list)
+        for user in query_results:
+            for role in user.roles or []:
+                for permission in role.oidc_permissions or []:
+                    if isinstance(permission, OidcPermissionAcsV1):
+                        permission_usernames[
+                            Permission(**permission.dict(by_alias=True))
+                        ].append(user.org_username)
+        return [
+            AcsRole.build(permission, usernames)
+            for permission, usernames in permission_usernames.items()
+        ]
+
+    def get_current_state(self, acs: AcsApi, auth_provider_id: str) -> list[AcsRole]:
+        """
+        Get current ACS roles and associated users from ACS api
+
+        :param acs: ACS api client
+        :param auth_id: id of auth provider within ACS instance to target for reconciliation
+        :return: list of current AcsRole associated with specified auth provider
+        """
+        current_roles: dict[str, AcsRole] = {}
+
+        roles = acs.get_roles()
+        groups = acs.get_groups()
+        role_assignments: RoleAssignments = self.build_role_assignments(
+            auth_provider_id, groups
+        )
+
+        for role in roles:
+            # process roles that are not system default
+            # OR
+            # system default roles referenced in auth rules
+            # however, do not reconcile the auth provider minimum access rule associated with 'None' system default
+            if not role.system_default or (
+                role.name in role_assignments and role.name != "None"
+            ):
+                try:
+                    access_scope = acs.get_access_scope_by_id(role.access_scope_id)
+                except Exception as e:
+                    logging.error(
+                        "Failed to retrieve current access scope: %s for role: %s\n\t%s",
+                        role.access_scope_id,
+                        role.name,
+                        e,
+                    )
+                    continue
+
+                try:
+                    permission_set = acs.get_permission_set_by_id(
+                        role.permission_set_id
+                    )
+                except Exception as e:
+                    logging.error(
+                        "Failed to retrieve current permission set: %s for role: %s\n\t%s",
+                        role.permission_set_id,
+                        role.name,
+                        e,
+                    )
+                    continue
+
+                current_roles[role.name] = AcsRole(
```

Comment:
`current_roles` is a `dict` and only values are returned, are we expecting there are duplicate `role.name` in `roles`, follow the current logic that last one wins, is this expected?

Reply:
Not expecting duplicates. Role name uniquess is enforced in ACS:
![image](https://github.com/app-sre/qontract-reconcile/assets/100163026/6c4a2779-e7d1-475c-95c4-0ebd02104aea)

Reply:
Then you can use `current_roles` as `list[AcsRole]`, do `current_roles.append(...` and return it directly.

Reply:
hmm. Yeah unsure why I thought usage of a dict was necessary here. Thanks
[updated](https://github.com/app-sre/qontract-reconcile/pull/3899/commits/aba611a97e850e5b8b93d32b91922c671cc62443)